core/app/models/spree/fulfilment_changer.rb

```diff
@@ -45,13 +45,64 @@ def handle_stock
           desired_stock_location.unstock(variant, unstock_quantity, desired_shipment)
         end
 
-        update_current_shipment_inventory_units(new_on_hand_quantity, :on_hand)
-        update_current_shipment_inventory_units(quantity - new_on_hand_quantity, :backordered)
+        move_inventory_units_between_shipments
+      end
+    end
+
+    def move_inventory_units_between_shipments
+      update_desired_shipment_inventory_units
+      update_current_shipment_inventory_units
+    end
+
+    def current_shipment_units
+      current_shipment.inventory_units.on_hand_or_backordered.where(variant_id: variant.id)
+    end
+
+    def update_desired_shipment_inventory_units
+      on_hand_unit = get_desired_shipment_inventory_unit(:on_hand)
+      on_hand_unit.update(quantity: on_hand_unit.quantity + new_on_hand_quantity)
+
+      new_backorder_quantity = current_on_hand_quantity - new_on_hand_quantity
+      if new_backorder_quantity > 0
+        backordered_unit = get_desired_shipment_inventory_unit(:backordered)
+        backordered_unit.update(quantity: backordered_unit.quantity + new_backorder_quantity)
+      end
+    end
+
+    def update_current_shipment_inventory_units
+      reduced_quantity = current_on_hand_quantity
+
+      # Reduce quantities from all backordered units first in case there is more than one.
+      backorder_units = current_shipment_units.backordered
```

Comment:
you are using this variable twice, maybe we could add here memoization ?

Reply:
@przemosk 👌 https://github.com/spree/spree/pull/10541/commits/d8022c7c820a7a77a6de03f6a52cbfaf6c4f330a